packages/auth/src/AuthProvider.tsx

```diff
@@ -1,34 +1,30 @@
 import React from 'react'
 
-import type {
-  AuthClient,
-  SupportedAuthTypes,
-  SupportedAuthClients,
-  SupportedUserMetadata,
-} from './authClients'
-import { createAuthClient } from './authClients'
-
 export interface CurrentUser {
   roles?: Array<string>
 }
 
-export interface AuthContextInterface {
+type PromiseValue<T> = T extends Promise<infer R> ? R : T
+
+export interface AuthContextInterface<WrappedClient extends AuthClient> {
   /* Determining your current authentication state */
   loading: boolean
   isAuthenticated: boolean
   /* The current user's data from the `getCurrentUser` function on the api side */
   currentUser: null | CurrentUser
   /* The user's metadata from the auth provider */
-  userMetadata: null | SupportedUserMetadata
-  logIn(options?: unknown): Promise<void>
-  logOut(options?: unknown): Promise<void>
-  signUp(options?: unknown): Promise<void>
+  userMetadata: null | PromiseValue<
+    ReturnType<WrappedClient['getUserMetadata']>
+  >
+  logIn: WrappedClient['logIn']
+  logOut: WrappedClient['logOut']
+  signUp: WrappedClient['signUp']
   getToken(): Promise<null | string>
   /**
    * Fetches the "currentUser" from the api side,
    * but does not update the current user state.
    **/
-  getCurrentUser(): Promise<null | CurrentUser>
+  getCurrentUser(): WrappedClient['getUserMetadata']
```

Comment:
My understanding is that `currentUser` and `userMetadata` are different things?
User metadata are returned by the `AuthClient` instance, whereas the "current user" data is returned by the API

Reply:
Oops, yeah, that was some overbearing typing :)